nixos/taler: init module

eljamm opened the pull request:
## Description of changes

[GNU Taler](https://taler.net/en/index.html) is an electronic payment system for privacy-preserving payments. For more details, see https://github.com/ngi-nix/projects/issues/1382.

Work funded by NLNet through the [NGI0 Entrust](https://nlnet.nl/entrust/) fund and packaged under the [Summer of Nix](https://github.com/ngi-nix/summer-of-nix) program.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Atemu committed:
nixos/taler: init module

basic config set that makes the service at least start

add secmod helpers and taler-global runtime dir

support for includes

taler denominations

Only enable services if taler is enabled

fix wirewatch service name

use correct permissions for database schema

The current permissions don't work or aren't enough and cause the
wirewatch and closer services to fail.

nixos/libeufin: init module

libeufin: refactor module

libeufin: add main service

nixos/taler: configure settings using options

Works, but can be refactored further

taler: refactor settings options

trim settings defaults to the absolutely necessary ones

nixos/libeufin: refactor and move to separate dir

nixos/libeufin: set defaultText

nixos/libeufin: use getExe

nixos/libeufin-bank: move to own dir

nixos/libeufin: move libeufin related config into its own config file

nixos/libeufin/bank: extract dbinitServiceName into var

nixos/libeufin: move script to ExecStart

nixos/libeufin: fix config file name

nixos/taler: refactor config file

nixos/taler-exchange: grant delete to taler-exchange-aggregator

Would repeatedly attempt to delete in a table where it wasn't allowed to and
cause insane spam in the postgres log.

nixos/taler/exchange: move exchange-specific options to exchange

nixos/taler: move generic taler settings into taler system module

nixos/taler: import exchange in module-list.nix

nixos/taler-exchange: refactor services group name

nixos/taler-exchange: use taler-harness to generate coins

The taler-wallet-cli does not have the deployment subcommand anymore,
but the docs still say that it should be used to generate the keys.

For now, the keys should be generated with taler-harness.

nixos/taler-exchange: add option to enable accounts

nixos/taler: add missing descriptions

nixos/taler(exchange): add description & use getExe'

nixos/taler(merchant): init submodule

nixos/taler: use correct script for db access

nixos/taler: merchant add depositcheck path

nixos/taler: review suggestions

nixos/taler: make runtimeDir into an option, refactor

nixos/taler: init mkTalerModule

nixos/taler: use mkTalerModule for exchange

nixos/taler: exchange fixups

nixos/taler: use mkTalerModule for merchant

nixos/taler: improve how dbInit script is created

nixos/taler: remove exchange enableAccounts option

nixos/taler: explicitly specify psql user

Sometimes the dbinit service fails to find the user.

nixos/taler: add openFirewall option; install package

feat: add assertions, remove throw

feat(taler): use module system instead of functions

Also:
- remove throw from denominateConfig
- rename `utils.nix` to `common.nix`

feat(taler): refactor modules

feat: move taler module to services/finance

refactor(exchange): replace throw with assert

refactor(exchange,merchant): settings options

fix(taler): manpage URLs

fix(exchange): public key assert

refactor(taler): use configFile

feat(taler): include component configs directly

Makes services detect config changes better.

eljamm committed:
tests/taler: init basic test

Test build and services start, but libeufin-bank fails when trying to
access the database to change the admin's password.

We need to execute the command as the libeufin-bank user.

tests/taler: add master private key

tests/taler: rewrite `register_bank_account` to Nix

tests/taler: rename libeufin node to bank

tests/taler: use xtaler wire_type instead of iban

tests/taler: remove redundant data from conf files

tests/taler: enable exchange account

tests/taler: remove unused talerConfig

tests/taler: add client node and attempt a withdrawal

tests/taler: systemd_run optional user and group args

tests/taler: refactor and make a withdrawal

tests/taler: refactor tasks into subtests

tests/taler: properly read and test balance

tests/taler: refactor commands and add comments

nixos/taler: rename private key

tests/taler: enable nexus service in bank node

tests/taler: nexus fake incoming payment test

tests/taler: use correct path for nexus client keys

tests/taler: add merchant node

tests/taler: merchant register instance

tests/taler: init pay for order merchant

tests/taler: fix payto uri

tests/taler: withdraw smaller amount

This makes the test faster

tests/taler: verify balance

tests/nixos: debugging merchant payment, cleanup

tests/taler: fix libeufin command, use curl to register accounts

tests/taler: add basic online test

tests/taler: move nodes into separate directory

tests/taler: fix insufficient balance error

Turns out that the exchange wire fees need to be set up (even if they're
0) in order for the CLI wallet to deposit coins into the merchant's bank
account.

tests/taler: improve node importing, port forwarding

tests/taler: import scripts from a separate file

tests/taler: move tests into a sub-directory

tests/taler: manually start services, cleanup

This results in less overhead and conflict since components will not try
to prematurely connect to the ones that haven't finished their set up.

tests/taler: remove online test

This was used to debug the insufficient balance problem, but it's not
really that useful by itself.

tests/taler: add nexus keys

tests/taler: use bank initalAccounts option

taler/tests: use initialAccount

tests/taler: make nexus work

tests/taler: don't run nexus test if there is no internet

tests/taler: use openFirewall, remove manual package install

fix(test): evaluation errors

fix(test): create nexus role by enabling createLocalDatabase

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,114 @@
+  buildInputs = [ nodePackages.nodejs ];
+
+  # Use a fake git?
+  postUnpack = ''
+    git init -b master
+    git config user.email "root@localhost"
+    git config user.name "root"
+    git commit --allow-empty -m "Initial commit"
+  '';
```
Why is this needed?

eljamm replied:
I suppose because the package does not build if it's not in a git dir. Here is what I get when I remove this:

```
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set)
```

dotlambda replied:
I suggest you try `touch .git`.

eljamm replied:
```
fatal: invalid gitfile format: /build/.git
```

eljamm replied:
Not only does the program want a git repo, but it also wants a commit to be present (perhaps it uses it for something?). So I don't think there is any way around doing this.

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,114 @@
+  srcOnly,
+}:
+let
+  nodePackages = nodejs.pkgs;
```
```suggestion
```

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,114 @@
+  nativeBuildInputs = [
+    git
+    jq
+    nodePackages.nodejs
```
```suggestion
    nodejs
```

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,114 @@
+    hash = "sha256-OwZLRv2ZWJ01pHWS+CGxvM4J19tGnEsRJ9S3bcsaWSE=";
+  };
+
+  buildInputs = [ nodePackages.nodejs ];
```
```suggestion
  buildInputs = [ nodejs ];
```

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,114 @@
+let
+  nodePackages = nodejs.pkgs;
+  nodeSources = srcOnly nodejs;
+  esbuild_0_19_9 = buildGoModule rec {
```
You can use
```suggestion
  esbuild_0_19_9 = esbuild.override {
```
instead. See e.g. the deltachat-desktop package.

dotlambda commented on pkgs/by-name/li/libeufin/package.nix:
```diff
@@ -0,0 +1,100 @@
+
+  meta = {
+    homepage = "https://git.taler.net/libeufin.git/";
+    description = "Integration and sandbox testing for FinTech APIs and data formats.";
```
```suggestion
    description = "Integration and sandbox testing for FinTech APIs and data formats";
```
https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

dotlambda commented on pkgs/by-name/li/libeufin/package.nix:
```diff
@@ -0,0 +1,100 @@
+    homepage = "https://git.taler.net/libeufin.git/";
+    description = "Integration and sandbox testing for FinTech APIs and data formats.";
+    license = lib.licenses.agpl3Plus;
+    maintainers = with lib.maintainers; [ atemu ];
```
Did @Atemu agree to this?

eljamm replied:
He was the one who put his name there, so I suppose so.

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,106 @@
+    homepage = "https://git.taler.net/wallet-core.git/";
+    description = "CLI wallet for GNU Taler written in TypeScript and Anastasis Web UI";
+    license = lib.licenses.gpl3Plus;
+    platforms = lib.platforms.linux;
```
mainProgram and maintainers missing

dotlambda commented on pkgs/by-name/ta/taler-wallet-core/package.nix:
```diff
@@ -0,0 +1,106 @@
+      (cd "$f" && (
+      npm run build-release --offline --nodedir="${nodeSources}"
+      find build -type f -exec \
+        ${removeReferencesTo}/bin/remove-references-to \
```
```suggestion
        ${lib.getExe removeReferencesTo} \
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-summer-of-nix-program-updates/46053/4

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+}:
+
+let
+  this = config.services.taler.exchange;
```
```
  cfg = config.services.taler.exchange;
```

eljamm replied:
I think this is a matter of preference. @Atemu suggested we use `this` instead of `cfg` and I don't really mind that.

SuperSandro2000 replied:
It is absolute standard to refer to the options defined in the same file with cfg unless there is a conflict or other things which make this not a good fit. I don't think we have that here.

asymmetric replied:
I agree with @SuperSandro2000, there's no other instance of `this` being used in such a way in nixpkgs AFAICT, using `cfg` is a very established idiom.

eljamm replied:
Alright, I'll rename it to `cfg`, then.

SuperSandro2000 commented on nixos/modules/services/taler/module.nix:
```diff
@@ -0,0 +1,66 @@
+in
+
+{
+  imports = [ ./exchange.nix ];
```
Please move this up into modules-list.nix

eljamm replied:
This has already been done (see [here](https://github.com/NixOS/nixpkgs/pull/332699/files#diff-7c411742b9908f3df53ed26638c28825bf5dc0886cd42f71e44c878783803829R1354))

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+  # taler-exchange needs a runtime dir shared between the taler services. Crypto
+  # helpers put their sockets here for instance and the httpd connects to them.
+  runtimeDir = "/run/taler-system-runtime/";
+  talerEnabled = config.services.taler.enable;
```
That should be replaced by cfgt.enable

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+  # helpers put their sockets here for instance and the httpd connects to them.
+  runtimeDir = "/run/taler-system-runtime/";
+  talerEnabled = config.services.taler.enable;
+  inherit (config.services.taler) configFile;
```
This is only used once, please inline it

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+    package = lib.mkPackageOption pkgs "taler-exchange" { };
+    denominationConfig = lib.mkOption {
+      type = lib.types.lines;
+      default = throw "You must set the denomination config `services.taler.exchange.denominationConfig`.";
```
We cannot use throw here

eljamm replied:
Why not? This is an important option which the program does not correctly work without.

The best way to mandate that the user sets it is this way.

SuperSandro2000 replied:
It would abort the evaluation and it is best to use assertions/warnings to get them all at once.

eljamm replied:
> it is best to use assertions/warnings to get them all at once

Is there an example of how I can do this?

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+      '';
+      description = ''
+        This option configures the cash denomination for the coins that the exchange offers.
+        For more information, consult the docs: https://docs.taler.net/taler-exchange-manual.html#coins-denomination-keys
```
```suggestion
        For more information consult the [upstream docs](https://docs.taler.net/taler-exchange-manual.html#coins-denomination-keys).
```

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+        serviceConfig = {
+          DynamicUser = true;
+          User = name;
+          Group = "taler-exchange"; # TODO refactor into constant
```
We can also just hardcode it

SuperSandro2000 commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,184 @@
+          User = name;
+          Group = "taler-exchange"; # TODO refactor into constant
+          ExecStart =
+            "${this.package}/bin/${name} -c ${configFile}" + lib.optionalString this.debug " -L debug"; # TODO as a list?
```
Please use lib.getExe. Also after ExecStart should be no new line

eljamm replied:
This is already the case ([here](https://github.com/NixOS/nixpkgs/pull/332699/files#diff-7c8ae3dbbd2b175f1e5d71e2fe0876f8e4d5b285c5681a2948d3ba5023d6b6c2R156)).

As for the new line, this is how nixfmt formats this. What do you suggest to do instead?

SuperSandro2000 commented on nixos/modules/services/taler/libeufin.nix:
```diff
@@ -0,0 +1,85 @@
+}:
+let
+  this = config.services.taler.libeufin.bank;
+  talerEnabled = config.services.taler.enable;
```
One of this must be cfg

SuperSandro2000 commented on nixos/modules/services/taler/libeufin.nix:
```diff
@@ -0,0 +1,85 @@
+    debug = lib.mkEnableOption "debug logging";
+  };
+
+  config = lib.mkIf (talerEnabled && this.enable) {
```
```suggestion
  config = lib.mkIf (cfgt.enable && cfg.enable) {
```

eljamm replied:
We're only using `this.enable` now ([here](https://github.com/NixOS/nixpkgs/pull/332699/files#diff-b0e1f52ea00abac3dcafdec2674afdee490e3e51c1084ee9b2aa5731b88b1690R27)).

SuperSandro2000 commented on nixos/modules/services/taler/libeufin.nix:
```diff
@@ -0,0 +1,85 @@
+    systemd.services = {
+      libeufin = {
+        script =
+          "${this.package}/bin/libeufin-bank serve -c ${configFile}"
```
Please use lib.getExe if possible

SuperSandro2000 commented on nixos/modules/services/taler/config.nix:
```diff
@@ -0,0 +1,143 @@
+}:
+
+let
+  this = config.services.taler;
```
Should be cfg

SuperSandro2000 commented on nixos/modules/services/taler/config.nix:
```diff
@@ -0,0 +1,143 @@
+      freeformType = settingsFormat.type;
+      options = {
+        # Should these be in here?
+        # Should these even be configurable?
```
Probably not, lets just hardcode them to make things easier for us

eljamm replied:
What do you think @Atemu ?

SuperSandro2000 commented on nixos/modules/services/taler/config.nix:
The changes in this commit done to the existing commits should be rolled up

SuperSandro2000 commented on nixos/modules/services/libeufin/module.nix:
```diff
@@ -0,0 +1,30 @@
+
+let
+  settingsFormat = pkgs.formats.ini { };
+  this = config.services.libeufin;
```
Should be done in cfg and be rolled up

eljamm replied:
What do you mean here exactly?

SuperSandro2000 replied:
Rolled up as in the commit should be merged with a previous one as it is a fixup commit.

SuperSandro2000 commented on nixos/tests/taler/basic.nix:
```diff
@@ -0,0 +1,261 @@
+            # https://git.taler.net/exchange.git/tree/src/testing/taler-unified-setup.sh?h=v0.11.2#n276
+
+            set -eux
+            wget \
```
we probably want to use curl and output it to stdout

SuperSandro2000 reviewed (commented):
I stopped after 2/3 because of the kinda messy history.

eljamm commented:
> I stopped after 2/3 because of the kinda messy history.

Yeah, the commit history isn't that great and old changes are being reviewed as a result.

I'll clean it up today.

eljamm committed:
nixos/libeufin: init module

nixos/libeufin: init module

nixos/libeufin(nexus): init submodule

nixos/libeufin(nexus): refactor service

Also added state directory to allow the creation of client ebic keys.

nixos/libeufin: review suggestions

nixos/libeufin: fix nexus service executable

nixos/libeufin: add mkLibeufinModule

nixos/libeufin: fix dbinit service not starting for utils, cleanup

nixos/libeufin: use mkLibeufinModule for nexus

nixos/libeufin: use mkLibeufinModule for bank

nixos/libeufin: add initialAccounts, stateDir options

nixos/libeufin: refactor to make nexus work, cleanup

nixos/libeufin: refactor stateDir, only register accounts on init

nixos/libeufin: explicitly specify psql user

Sometimes the dbinit service fails to find the user.

nixos/libeufin: cleanup stateDir

nixos/libeufin: add openFirewall option; install package

feat: apply review suggestions

Co-authored-by: h7x4 <h7x4@nani.wtf>

style: format code

fix: evaluation errors

fix(libeufin): start main services after dbinit

The main services can start after their databases have been initialized,
it's just that the bank and nexus shouldn't do the initialization at the
same time.

refactor(libeufin): dbinit script

feat: add assertions, remove throw

chore: remove unused code

feat(libeufin): recfactor dbinit service

feat: move libeufin module to services/finance

refactor(libeufin): remove configFile option

refactor(libeufin): use environment.etc for config file

eljamm commented:
The history should be better now.

eljamm commented:
Rebased on top of master to include the [taler: 0.12 -> 0.13](https://github.com/NixOS/nixpkgs/pull/338184) update.

No tests have been broken from this as breaking API changes have already been fixed beforehand.

SuperSandro2000 commented:
I tried to review the changes commit by commit and got lead into some dead ends through that. 😅 You said you fixed that which should make things easier but I currently lack some motivation to review this big PR again.

asymmetric commented on nixos/modules/services/libeufin/bank.nix:
```diff
@@ -0,0 +1,121 @@
+        serviceConfig = {
+          DynamicUser = true;
+          User = bankServiceName;
+          ExecStart =
+            "${lib.getExe this.package} serve -c ${configFile}" + lib.optionalString this.debug " -L debug";
+        };
```
IMO this looks a bit nicer:

```suggestion
          ExecStart = toString [
            (lib.getExe this.package)
            "serve -c ${configFile}"
            (lib.optionalString this.debug "-L debug")
          ];
        };
```

eljamm replied:
Yeah, I guess it's a bit cleaner like this.

asymmetric commented on nixos/modules/services/libeufin/bank.nix:
```diff
@@ -0,0 +1,121 @@
+      };
+      ${dbinitServiceName} = {
+        path = [ config.services.postgresql.package ];
+        script =
```
Why `script` instead of `ExecStart`?

eljamm replied:
You're right, an `ExecStart` is enough, here.

asymmetric commented on nixos/modules/services/libeufin/nexus.nix:
```diff
@@ -0,0 +1,172 @@
+            User = serviceName;
+            ExecStart = "${nexusExe} serve -c ${configFile}" + lib.optionalString this.debug " -L debug";
+            StateDirectory = serviceName;
+            ReadWritePaths = [ "/var/lib/${serviceName}" ];
```
Would it make sense to make this configurable as a `dataDir` option, instead of hard-coding `/var/lib`? See e.g. the [synapse](
https://github.com/NixOS/nixpkgs/tree/717ca7bea5dc1017be222fbba349931b4286786d/nixos/modules/services/matrix/synapse.nix#L1235) moduie. You'd have to ensure it's created though.

eljamm replied:
Yes, this does make sense. Thanks for the suggestion.

eljamm replied:
> You'd have to ensure it's created though.

I think this is already handled by systemd.

asymmetric commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,242 @@
+      default = { };
+    };
+    enableAccounts = lib.mkOption {
+      type = lib.types.listOf lib.types.path;
```
You used `with` in such a case elsewhere, you might want to do it consistently

```suggestion
      type = with lib.types; listOf path;
```

eljamm replied:
There is one instance of `with lib` in `nixos/modules/services/taler/module.nix`. I'll remove it.

asymmetric commented on nixos/modules/services/taler/exchange.nix:
```diff
@@ -0,0 +1,242 @@
+        };
+      };
+
+    users.groups.${groupName} = { };
```
This might be unnecessary since you use `DynamicUser`. From `man systemd.exec`:

> When used in conjunction with DynamicUser= the user/group name specified is dynamically allocated at the time the service is started, and released at the time the service is stopped — unless it is already allocated statically (see below). If DynamicUser= is not used the specified user and group must have been created statically in the user database no later than the moment the service is started, for example using the sysusers.d(5) facility, which is applied at boot or package install time. If the user does exist by then program invocation will fail.

Or do you need the group before the service is started/after it's stopped?

asymmetric reviewed (commented):
Did my best to review the "interesting" bits.

eljamm commented:
Thanks for reviewing @asymmetric. Please let me know if I missed anything.

asymmetric commented:
(@eljamm I've reached out on matrix re: the work in this PR FYI :+1: )

eljamm commented:
I'll keep the changes as separate commits and then merge them together when the commit history gets too long. Hopefully this makes reviewing easier.

One notable change here is the introduction of `mkTalerModule` to set up some common things between modules, which should reduce some redundancy and make the code easier to read.

eljamm commented:
I think the module has reached a good state to be merged as a first iteration as its components' core functionality has been successfully tested, which can be verified by running:

```sh
nix run .#nixosTests.taler.basic.driver
```

This is the first time I handle a big PR like this on a non-personal level so I apologize if this has been hard to review. On that note, I'd appreciate any pointers or suggestions on how I can improve this going forward.

eljamm commented:
Added `openFirewall` option and packages are now installed in the system when the modules are enabled.

h7x4 commented:
Could you move the modules into a fitting subcategory? I think `nixos/modules/services/finance` could work, alternatively there's `nixos/modules/services/misc`

h7x4 commented on nixos/modules/services/libeufin/common.nix:
```diff
@@ -0,0 +1,161 @@
+          after = [ "${serviceName}-dbinit.service" ];
+          wantedBy = [ "multi-user.target" ];
+
+          # TODO: Could this preStart block be part of the db-init service, and then
+          #       we add it as a dependency? It seems to be part of database initialization
+          #       in a certain sense
+
+          # Some accounts might need to be registered before the
+          # service starts, like the exchange when the bank's currency
+          # conversion is enabled.
+          preStart =
+            let
+              initialAccountRegistration = lib.concatMapStringsSep "\n" (
+                account:
+                let
+                  args = lib.cli.toGNUCommandLineShell { } {
+                    c = cfgMain.configFile;
+                    inherit (account) username password name;
+                    payto_uri = "payto://x-taler-bank/bank:${toString bankPort}/${account.username}?receiver-name=${account.name}";
+                    exchange = lib.toLower account.username == "exchange";
+                  };
+                in
+                "${lib.getExe' cfg.package "libeufin-bank"} create-account ${args}"
+              ) cfg.initialAccounts;
+            in
+            lib.mkIf ((!isNexus) && (cfg.initialAccounts != [ ])) ''
+              # only register initial accounts once
+              if [ ! -e /var/lib/${serviceName}/init ]; then
+                ${initialAccountRegistration}
+
+                touch /var/lib/${serviceName}/init
+                echo "Bank initialisation complete"
+              fi
+            '';
```
Would it make sense to move this block to `"${serviceName}-dbinit"`, so all of the database initialization is collected into one place?

eljamm replied:
It does make sense, but in that case the service should always start, even when both databases are remote and no initialization happens.

h7x4 replied:
I'm not sure I understand your answer correctly. The `requires = [ "${serviceName}-dbinit.service" ]` and `after = [ "${serviceName}-dbinit.service" ]` should ensure that the services only can run when the database initialization has taken place and exits successfully.

It's a bit awkward to store a file on a local server stating that a remote database has been initialized, but that's a separate issue.

eljamm replied:
> The requires = [ "${serviceName}-dbinit.service" ] and after = [ "${serviceName}-dbinit.service" ] should ensure that the services only can run when the database initialization has taken place and exits successfully.

Yes, but what I meant is that if we use remote databases for the bank and nexus, the dbinit service doesn't start since it depends on the postgresql service, right?

```nix
serviceDependencies =
  # The bank and nexus shouldn't initialise the database at the same time
  # This is only enforcable if both services are running on the same machine
  (lib.optionals (isNexus && cfgMain.bank.enable) [ "libeufin-bank-dbinit.service" ])
  ++ (lib.optionals cfg.createLocalDatabase [ "postgresql.service" ]);
```

That's why I changed it to this:

```nix
requires = lib.optionals cfg.createLocalDatabase [ "postgresql.service" ];
after = [ "network.target" ] ++ lib.optionals cfg.createLocalDatabase [ "postgresql.service" ];
```

**PS:** database initialization is now done from the same service, which is much simpler.

> It's a bit awkward to store a file on a local server stating that a remote database has been initialized, but that's a separate issue.

Yeah, but I didn't find any better solution than this.

h7x4 replied:
> [...] dbinit service doesn't start since it depends on the postgresql service, right?

As long as the dependencies of the `dbinit` service is set correctly (which they are with the `createLocalDatabase` conditionals), I think everything should be okay. The requires from the normal service unit will ensure this one is started beforehand as well.

> **PS:** database initialization is now done from the same service, which is much simpler.

Great, that was what I was asking for!

h7x4 commented on nixos/modules/services/libeufin/common.nix:
```diff
@@ -0,0 +1,161 @@
+    enable = lib.mkEnableOption "libeufin core banking system and web interface";
+    package = lib.mkPackageOption pkgs "libeufin" { };
+    debug = lib.mkEnableOption "debug logging";
+    # TODO: find a better way to
+    openFirewall = lib.mkEnableOption "Open ports in the firewall";
```
This is an incomprehensible comment left over from my review patch. What I meant is that the resulting doccomment for `openFirewall` is `Whether to Open ports in the firewall.`. I was thinking maybe it should be worded better. At the very least the first letter should be lowercased.

h7x4 commented on nixos/modules/services/libeufin/common.nix:
```diff
@@ -0,0 +1,161 @@
+      isNexus = libeufinComponent == "nexus";
+
+      # get database name from config
+      # TODO: should this always be the same db? In which case, should this be an option directly under `services.libeufin`?
+      dbName =
+        lib.removePrefix "postgresql:///"
+          cfg.settings."libeufin-${libeufinComponent}db-postgres".CONFIG;
```
I'll just leave a comment here, to mark it as a point of discussion. If `nexus` and `bank` always uses the same database url and credentials, should we configure it directly under `services.libeufin`?

eljamm replied:
Yeah, I think this might be necessary, unless there is a way to give permissions across databases.

h7x4 commented on nixos/modules/services/taler/module.nix:
```diff
@@ -0,0 +1,108 @@
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+        options = {
+          # Should these be in here?
+          # Should these even be configurable?
+          PATHS = {
+            TALER_DATA_HOME = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = "\${STATE_DIRECTORY}/";
+            };
+            TALER_CACHE_HOME = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = "\${CACHE_DIRECTORY}/";
+            };
+            TALER_RUNTIME_DIR = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = cfg.runtimeDir;
+            };
+          };
```
Maybe these could be set with these default values in `config` instead? It might make the code less awkward. Then people can override them on their own.

```nix
config = lib.mkIf cfg.enable {
  ...
  services.taler.settings.PATHS = {
    TALER_DATA_HOME = lib.mkDefault "\${STATE_DIRECTORY}/";
    TALER_CACHE_HOME = lib.mkDefault "\${CACHE_DIRECTORY}/";
    TALER_RUNTIME_DIR = cfg.runtimeDir;
  };
};
```

eljamm replied:
I don't know if normal users should change these options as functionality can break if they're misconfigured.

h7x4 replied:
I'm a bit wary of setting things so that the upstream users cannot modify them whatsoever. People tend to have weird usecases for things that I hadn't thought about before. Like for example, someone might want to store the `DATA_HOME` on a different drive and for some reason can't resort to bindmounts, or want to store their cache in an in-memory disk.

If you remove the `lib.mkDefault`, people would have to use `mkForce` as if to say "I know what I'm doing". How does that sound?

h7x4 commented on nixos/modules/services/taler/module.nix:
```diff
@@ -0,0 +1,108 @@
+            CURRENCY_ROUND_UNIT = lib.mkOption {
+              type = lib.types.str;
+              default = "${cfg.settings.taler.CURRENCY}:0.01";
+              defaultText = "0.01 in {option}`CURRENCY`";
```
```suggestion
              defaultText = lib.literalExpression ''
                "''${config.services.taler.settings.taler.CURRENCY}:0.01"
              '';
```

h7x4 commented on nixos/modules/services/taler/module.nix:
```diff
@@ -0,0 +1,108 @@
+        This allows including arbitrary INI files, including imperatively managed ones.
+      '';
+    };
+    configFile = lib.mkOption {
+      internal = true;
+      default =
+        let
+          includes = pkgs.writers.writeText "includes.conf" (
+            lib.concatStringsSep "\n" (map (include: "@inline@ ${include}") cfg.includes)
+          );
+          generatedConfig = settingsFormat.generate "generated-taler.conf" cfg.settings;
+        in
+        pkgs.runCommand "taler.conf" { } ''
+          cat ${includes} > $out
+          echo >> $out
+          echo >> $out
+          cat ${generatedConfig} >> $out
+        '';
+    };
```
I think it would be better if this was taken out of `options` and passed to `common.nix` via other means, since it doesn't look like something the module user should be touching directly. Since you're already adding it to `environment.etc`, maybe you could reference this value in `common.nix` instead?

eljamm replied:
This option defaults to the generated config file, but also allows the user to use an external one. Wouldn't this be useful?

Atemu commented on nixos/modules/services/taler/module.nix:
```diff
@@ -0,0 +1,108 @@
+          };
+          taler = {
+            CURRENCY = lib.mkOption {
+              type = lib.types.str;
+              defaultText = "None, you must set this yourself.";
```
The defaultText was intentional set to explicitly inform the user that they must set this setting.

h7x4 replied:
Yes, and I'm suggesting we do not do that. There are probably thousands of options around nixpkgs at this point that have this requirement, and it's better if we use the module system, option `example`s, and `meta.doc`s to tell users what they need to set to have a working service. The module system could definitely throw a better error message when something required is left unset, but I do not think that using `defaultText` for this unintended purpose is a good solution.

h7x4 commented on nixos/modules/services/libeufin/common.nix:
```diff
@@ -11,8 +11,7 @@ libeufinComponent:
     debug = lib.mkEnableOption "debug logging";
-    # TODO: find a better way to
-    openFirewall = lib.mkEnableOption "Open ports in the firewall";
+    openFirewall = lib.mkEnableOption "whether to open ports in the firewall";
```
Sorry, I don't think I was clear enough with this one. `mkEnableOption` is a bit confusing, because it adds your string in the middle of a `Whether to enable ...` block

https://github.com/NixOS/nixpkgs/blob/e9e2212a0715e5aa36fdbfb14f80b5a4508cf3c8/lib/options.nix#L91-L105

It needs to either be reworded, overwritten with `lib.mkEnableOption "" // { description = "..."; }` or just use a normal `lib.mkOption`.

h7x4 commented on nixos/modules/services/libeufin/common.nix:
```diff
@@ -58,54 +49,42 @@ libeufinComponent:
 
         # Database Initialisation
-        "${serviceName}-dbinit" =
+        libeufin-dbinit =
```
Will this now be defined twice, with different contents for both nexus and bank?

eljamm replied:
The contents will be appended in one single service `libeufin-dbinit`. For example, the start script will initialize the bank, then the nexus:

```sh
#!/nix/store/717iy55ncqs0wmhdkwc5fg2vci5wbmq8-bash-5.2p32/bin/bash
set -e

/nix/store/4cvdvyaqpvcla1m7ks6n95ln33hncn4h-libeufin-0.13.0/bin/libeufin-bank dbinit -L debug -c /nix/store/lfhm2saxf9vjyrrdr5ni9p7pvg8abmym-generated-libeufin.conf

/nix/store/4cvdvyaqpvcla1m7ks6n95ln33hncn4h-libeufin-0.13.0/bin/libeufin-nexus dbinit -L debug -c /nix/store/lfhm2saxf9vjyrrdr5ni9p7pvg8abmym-generated-libeufin.conf
```

That's why we don't have to worry about them running at the same time and conflicting anymore.

eljamm commented:
I moved the modules to `services/finance`, which I forgot to do before.

I also don't know if there any more suggestions, but I think this PR is in a good state to be merged as a first iteration, which we can build upon and improve later on.

asymmetric commented on nixos/tests/taler/common/nodes.nix:
```diff
@@ -0,0 +1,200 @@
+                # == Mandatory ==
+                CURRENCY = FIAT_CURRENCY;
+                # Bank
+                HOST_BASE_URL = "https://isotest.postfinance.ch/ebicsweb/ebicsweb";
```
This means the test at line 226 doesn't run in this suite -- is that the idea? Would it be possible to run a service as part of the test and point `HOST_BASE_URL` there?

eljamm replied:
> This means the test at [line 231](https://github.com/NixOS/nixpkgs/blob/9425b7cc6895eba722c1080b35addc6992d38496/nixos/tests/taler/tests/basic.nix#L231) doesn't run in this suite, since the curl to an external URL fails -- is that the idea?

That's correct. If we can connect to the PostFinance platform, then we run the Nexus' currency conversion test, else we skip it. This was done because the NixOS test doesn't have access to the internet unless it's ran interactively, which is indicated with [a message](https://github.com/NixOS/nixpkgs/blob/9425b7cc6895eba722c1080b35addc6992d38496/nixos/tests/taler/tests/basic.nix#L7-L12) at the beginning.

> Would it be possible to run a service as part of the test and point `HOST_BASE_URL` there?

Only if we can run an EBICS/PostFinance server, which I don't think exists in nixpkgs, yet. I also don't know if any self-hostable solutions exist either or how much effort it is to package them it they do.

asymmetric reviewed (commented):
I ran the tests and besides the expected skipped tests, everything went fine.

I'm leaning towards merging this once my comment is addressed, as this has been open forr a long time and we can iterate on it in smaller chunks afterwards.

dali99 commented:
I'd like to test this properly before it's merged. I have an in-progress test-deployment, and the module has only been in a "looks good overall" state for ~2 weeks.

I'm not sure how much it has been tested by others, so maybe the interfaces are good, but if we're not sure I think it's better to miss the release cycle, and not have to mess it up too much with `mkRemovedOptionModule` and such if any UX improvements would be wanted otherwise.

fricklerhandwerk commented:
@dali99 any objections to merging? This will land in unstable anyway now.

dali99 commented on nixos/modules/services/finance/taler/exchange.nix:
```diff
@@ -0,0 +1,148 @@
+
+{
+  imports = [
+    (import ./common.nix { inherit talerComponent servicesDB servicesNoDB; })
```
When I tried to do something similar to this in the matrix-synapse module, both @SuperSandro2000 and @Ma27 got the heebie-jeebies.

I'm not really sure of a much better solution, but I think generally people might have some apprehensions about not importing all the things through the module list.

I still run my synapse module with this, and its been fine for years, so I dont mind, but I'm calling attention to this in case anyone else has feedback here/missed something.

asymmetric replied:
@dali99 do you have a link to the PR, for more context?

dali99 replied:
The relevant links are https://github.com/NixOS/nixpkgs/pull/242026/files#r1255517843 and https://github.com/NixOS/nixpkgs/pull/241973#pullrequestreview-1618774173

There are no specific points made about the dangers or anything, so /shrug

dali99 replied:
One difference from here might revolve around how I passed in `cfg.package.python`. Which depends on the module system itself (and thus maybe evaluation order?). The Taler module in question here doesn't do that (passing in just normal variables), so maybe it's less of a problem.

dali99 commented on nixos/modules/services/finance/taler/exchange.nix:
```diff
@@ -0,0 +1,148 @@
+          };
+        };
+      };
+      default = { };
```
I think adding a full example to the `services.taler.${component}.settings.${component}` would be a big value add.
The nixos manual only kind of half-documents these things as it stands.
So sneaking in a full example somewhere is nice.

Either here or in `services.taler.${component}.settings` maybe?

dali99 commented on nixos/modules/services/finance/libeufin/module.nix:
```diff
@@ -0,0 +1,29 @@
+
+{
+  options.services.libeufin = {
+    configFile = lib.mkOption {
+      internal = true;
+      default = settingsFormat.generate "generated-libeufin.conf" cfg.settings;
+    };
```
Commonly one would put this in a let-block.

```
let
  cfg = config.services.libeufin;
  format = pkgs.formats.ini { };
  configFile = format.generate "generated-libeufin.conf" cfg.settings;
in
```

Using the module system and `internal` to store variables like this is a pattern readers would immediately get very suspicious about.

dali99 commented on nixos/modules/services/finance/taler/merchant.nix:
```diff
@@ -0,0 +1,109 @@
+        options = {
+          # TODO: do we want this to be a sub-attribute or only define the merchant set of options here
+          merchant = {
+            DB = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = "postgres";
+              description = "Plugin to use for the database.";
+            };
```
Another option, if you dont want users to set this. Is to just configure it in `config`.

```
config.services.taler.merchant.settings.DB = "postgres";
```

This will just set the option with a high priority level, and if a user tries to  override it will error out unless they use `lib.mkForce` or similar. I think that's what you want here.

dali99 commented on nixos/modules/services/finance/taler/merchant.nix:
```diff
@@ -0,0 +1,109 @@
+                Whether the HTTP server should listen on a UNIX domain socket ("unix") or on a TCP socket ("tcp").
+              '';
+            };
+            LEGAL_PRESERVATION = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = "10 years";
+              description = "How long to keep data in the database for tax audits after the transaction has completed.";
+            };
```
Why is this one internal?

eljamm replied:
In retrospect, this might have been a bit unnecessary as this is part of the taler settings and not something for NixOS devs, so I removed it earlier.

Currently, the only internal option is `settings.services.taler.merchant.merchantdb-postgres.SQL_DIR`, which I think should stay that way since it points to the nix store.

D4ndellion reviewed (changes requested):
[I did manage to deploy it, based on the test](https://github.com/Programvareverkstedet/pvv-nixos-config/tree/892e3e4d6f1aa95561d8f7416101b0b6d8720eed/hosts/kvernberg/services/pvvvvvv), and I think the overall structure of the options is good! Great job! Most of the weird structure comes from taler itself.

There are some practical issues, like some of the tooling not working out of the box. (#357535) is one change to address this.

For future work, I think some manual section that details the imperative steps required to get it running would be nice.
The upstream docs could be difficult for someone to "translate" without being pretty well-versed with NixOS.

There's also a lot of - uncommon - patterns around. I commented on some of it, but a lot of the usage of `internal` especially seem very non-standard/non-obvious. I would like some explanation around why this is used so much (maybe I missed somewhere in the thread, though. If so; apologies!).

I would consider the systemd unit restarting stuff at least a blocker for merge

I haven't tried deploying libeufin-nexus, or the merchang backend, so I can't comment too much on that.

And of course, overall good job! Thanks for doing this!

eljamm commented on nixos/modules/services/finance/taler/common.nix:
```diff
@@ -0,0 +1,117 @@
+        };
+        requires = [ "taler-${talerComponent}-dbinit.service" ];
+        after = [ "taler-${talerComponent}-dbinit.service" ];
+        wantedBy = [ "multi-user.target" ]; # TODO slice?
+      }))
```
I've changed it so the main config uses the generated components configs instead of the `/etc/` paths,, so any changes to the components will change the main config, effectively causing the services to restart.

dali99 commented on nixos/modules/services/finance/taler/common.nix:
```diff
@@ -0,0 +1,117 @@
+      allowedTCPPorts = [ cfg.settings."${talerComponent}".PORT ];
+    };
+
+    environment.systemPackages = [ cfg.package ];
```
`propagatedBuildInputs` won't patch the scripts in `$out/bin`, which include

* `taler-config` (uses gnunet-config)
* `taler-terms-generator` (uses taler-config)
* `taler-exchange-dbconfig` (uses taler-config)
* `taler-auditor-dbconfig`(uses taler-config)
* `taler-exchange-helper-measure-test-form` (uses taler-config)
* `taler-unified-setup.sh` (uses taler-config)

I dont know if the `dbconfig` ones are important, and the `taler-terms-generator` is blocked on https://github.com/NixOS/nixpkgs/pull/357535 anyways. (though I would consider that one kind of important)

asymmetric reviewed (commented):
@eljamm can any of these tests be addded as a `passthru.tests` to one of the taler packages? Or are they E2E tests and therefore touch on several components?

eljamm commented:
> @eljamm can any of these tests be addded as a passthru.tests to one of the taler packages? Or are they E2E tests and therefore touch on several components?

In my opinion, separate component tests don't mean much right now. For example, we can make a test where we set up the exchange, but that won't tell us that it works correctly. For that, we need the other components, and that's what the basic test basically does.

That said, we can add the basic test to each package's `passthru.tests`, which will tell us when the workflow breaks (on major API changes, for example). Don't know if there is something better we can do at the moment.

fricklerhandwerk commented:
Merging now, since there were no further issues or objections. @eljamm is on our NGI budget to fix forward if any problems arise.
